### PR TITLE
basic: fix ASSERT_SIDE_EFFECT Coverity report in assert_se()

### PR DESCRIPTION
**Pull request by mrc0mmand**
As I, once again, stumbled upon the dreaded `ASSERT_SIDE_EFFECT` Coverity report, the respective [ticket](https://community.synopsys.com/s/question/0D534000046Yuzb/suppressing-assertsideeffect-for-functions-that-allow-for-sideeffects) and I also got my hands on the Coverity utilities, I tried to *delve* a little bit deeper and it appears that a certain combination of solutions from the aforementioned tickets actually works.

Reproducer:
```bash
meson build
mkdir coverity
cov-configure --config coverity/coverity_config.xml --template --compiler cc --comptype gcc
ninja -C build clean
cov-build --config coverity/coverity_config.xml --dir coverity/intermediate --force ninja -C build/
cov-analyze --dir coverity/intermediate/ --disable-default --enable ASSERT_SIDE_EFFECT --strip-path `pwd`
```

Current master:
```
Analysis summary report:
------------------------
Files analyzed                 : 423
Total LoC input to cov-analyze : 481031
Functions analyzed             : 3676
Time taken by analysis         : 00:00:03
Defect occurrences found       : 178 ASSERT_SIDE_EFFECT
```

With this patch:
```
Analysis summary report:
------------------------
Files analyzed                 : 423
Total LoC input to cov-analyze : 481031
Functions analyzed             : 3676
Time taken by analysis         : 00:00:03
Defect occurrences found       : 0
```

### CODE REVIEW

**Comment by mrc0mmand** [comment]
Okay, after a little bit of playing around I think I suspect why *Stefan* (from the Coverity ticket) couldn't reproduce the issue and why this fix is probably incomplete.

The fix seems to work alright, until you see the number of files actually processed by the local Coverity utilities:

```
# cov-build --config coverity/coverity_config.xml --dir coverity/intermediate --force -- ninja -C build/ 
Coverity Build Capture (64-bit) version 2019.09 on Linux 5.0.16-300.fc30.x86_64 x86_64
...
[WARNING] Emitted 2 C/C++ compilation units (0%) successfully

378 C/C++ compilation units (38%) are ready for analysis
 For more details, please look at: 
    /root/systemd/coverity/intermediate/build-log.txt
```

With this *subset* of files (and this patch), the `cov-analyze` command gives the output mentioned in the previous comment:

```
Analysis summary report:
------------------------
Files analyzed                 : 423
Total LoC input to cov-analyze : 481031
Functions analyzed             : 3677
Time taken by analysis         : 00:00:03
Defect occurrences found       : 0
```

However, if you look into the build log to inspect the suspiciously low number of processed files (38%), you'll find this:

```
[16558] CWD: /root/systemd/build
[16558] COMPILING: /opt/cov-sa-2019.09/bin/cov-translate cc -o fuzz-udev-database fuzz-udev-database@exe/src_fuzz_fuzz-udev-database.c.o fuzz-udev-database@exe/src_fuzz_fuzz-main.c.o -Wl,--no-undefined -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--start-group src/shared/libsystemd-shared-243.so -Wl,--end-group -Wl,-rpath,$ORIGIN/src/shared -Wl,-rpath-link,/root/systemd/build/src/shared
COV_TRANSLATE EXPANDED ARGS: "-o" "fuzz-udev-rules" "fuzz-udev-rules@exe/src_fuzz_fuzz-udev-rules.c.o" "fuzz-udev-rules@exe/src_fuzz_fuzz-main.c.o" "-Wl,--no-undefined" "-Wl,--as-needed" "-Wl,-z,relro" "-Wl,-z,now" "-Wl,--start-group" "src/udev/libudev-core.a" "src/shared/libsystemd-shared-243.so" "src/udev/libudev_static.a" "src/udev/libudev-basic.a" "src/libsystemd-network/libsystemd-network.a" "-lacl" "/usr/lib64/libblkid.so" "/usr/lib64/libkmod.so" "-Wl,--end-group" "-pthread" "-Wl,-rpath,$ORIGIN/src/udev:$ORIGIN/src/shared:$ORIGIN/src/libsystemd-network" "-Wl,-rpath-link,/root/systemd/build/src/udev:/root/systemd/build/src/shared:/root/systemd/build/src/libsystemd-network"
"../src/systemd/_sd-common.h", line 23: error #35: #error directive: "Do not
          include _sd-common.h directly; it is a private header."
  #  error "Do not include _sd-common.h directly; it is a private header."
```

After *solving* this temporarily with this pretty ugly patch:

```diff
diff --git a/src/systemd/_sd-common.h b/src/systemd/_sd-common.h
index b3ee7bbc24..cd3651f835 100644
--- a/src/systemd/_sd-common.h
+++ b/src/systemd/_sd-common.h
@@ -19,9 +19,6 @@
 
 /* This is a private header; never even think of including this directly! */
 
-#if defined(__INCLUDE_LEVEL__) && __INCLUDE_LEVEL__ <= 1
-#  error "Do not include _sd-common.h directly; it is a private header."
-#endif
 
 typedef void (*_sd_destroy_t)(void *userdata);
 
```

The `cov-build` utility now yields this output:

```
# cov-build --config coverity/coverity_config.xml --dir coverity/intermediate --force -- ninja -C build/ 
Coverity Build Capture (64-bit) version 2019.09 on Linux 5.0.16-300.fc30.x86_64 x86_64
...
Emitted 616 C/C++ compilation units (100%) successfully
[WARNING] Recoverable errors were encountered during 2 of these C/C++ compilation units.

992 C/C++ compilation units (100%) are ready for analysis
 For more details, please look at: 
    /root/systemd/coverity/intermediate/build-log.txt
```

accompanied by following report from `cov-analyze`:

```
Analysis summary report:
------------------------
Files analyzed                 : 1134
Total LoC input to cov-analyze : 490722
Functions analyzed             : 14073
Time taken by analysis         : 00:00:06
Defect occurrences found       : 101 ASSERT_SIDE_EFFECT
```

All those *defects* are only from the code under `src/libsystemd`:
```
$ cd coverity/intermediate/output
# grep file ASSERT_SIDE_EFFECT.errors.xml | wc -l
101
# grep file ASSERT_SIDE_EFFECT.errors.xml | grep libsystemd | wc -l
101
```

However, all those reports mention macro `assert_return` - see https://mrc0mmand.fedorapeople.org/systemd_coverity_assert/ - which is weird, as all reports are only from sources under `src/libsystemd`, although the `assert_return` macro is used throughout the entire codebase.

I'm not sure if I'm missing something obvious, but this whole situation is somewhat confusing.

**Comment by mrc0mmand** [comment]
Regarding the `_sd-common.h` include issue, looking at, for example, Tor project, they apparently had a similar issue, judging from this piece of code (scattered on multiple places):

```C
#ifdef HAVE_SYSTEMD
#  if defined(__COVERITY__) && !defined(__INCLUDE_LEVEL__)
/* Systemd's use of gcc's __INCLUDE_LEVEL__ extension macro appears to confuse
 * Coverity. Here's a kludge to unconfuse it.
 */
#   define __INCLUDE_LEVEL__ 2
#endif /* defined(__COVERITY__) && !defined(__INCLUDE_LEVEL__) */
#include <systemd/sd-daemon.h>
#endif /* defined(HAVE_SYSTEMD) */
```

https://github.com/torproject/tor/search?q=__INCLUDE_LEVEL__&unscoped_q=__INCLUDE_LEVEL__

**Comment by filbranden** [comment]
Ah great that you found about `_sd-common.h`! I think that somewhat explains why in the past I thought we succeeded in this and then later it was clear we didn't...

Maybe we might want to include the Tor snippet in our sources, at least to help fix that particular issue with Coverity.

Using the Coverity for OSS projects we have available (scan.coverity.com) makes it fairly hard to troubleshoot this kind of issues... If you have access to a license of Coverity and some Coverity experts, maybe see if they can help uncover more issues here as well.

Cheers!
Filipe

**Comment by mrc0mmand** [comment]
I had a little chat with our *internal* Coverity guru and, unfortunately, there doesn't seem to be any viable solution, so far. We solved this in our internal Coverity instance by patching the tools to ignore `assert_se` and `assert_return` macros, however, we can't do that in upstream.

File: src/basic/macro.h [hunk 1]
```diff
@@ -307,17 +307,18 @@ static inline unsigned long ALIGN_POWER2(unsigned long u) {
 
 extern void __coverity_panic__(void);
 
-static inline int __coverity_check__(int condition) {
+static inline void __coverity_check__(int condition) {
+        if(!condition)
```
Comment (keszybz) [inline src/basic/macro.h]: Looks good, but please add a space after the if.

Comment (mrc0mmand) [reply src/basic/macro.h]: Whoops, missed this one, thanks!

**Pull request merged**